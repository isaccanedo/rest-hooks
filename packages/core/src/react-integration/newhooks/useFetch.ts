/* eslint-disable @typescript-eslint/ban-ts-comment */
import { ExpiryStatus } from '@rest-hooks/normalizr';
import {
  EndpointInterface,
  Denormalize,
  Schema,
  FetchFunction,
} from '@rest-hooks/normalizr';
import { useMemo } from 'react';

import useController from '../hooks/useController.js';
import useCacheState from './useCacheState.js';

/**
 * Request a resource if it is not in cache.
 * @see https://resthooks.io/docs/api/useFetch
 */
export default function useFetch<
  E extends EndpointInterface<FetchFunction, Schema | undefined, undefined>,
  Args extends readonly [...Parameters<E>] | readonly [null],
>(endpoint: E, ...args: Args) {
  const state = useCacheState();
  const controller = useController();

  const key = args[0] !== null ? endpoint.key(...args) : '';
  const cacheResults = key && state.results[key];
  const meta = state.meta[key];

  // Compute denormalized value
  const { expiryStatus, expiresAt } = useMemo(() => {
    // @ts-ignore
    return controller.getResponse(endpoint, ...args, state) as {
      data: Denormalize<E['schema']>;
      expiryStatus: ExpiryStatus;
      expiresAt: number;
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [
    cacheResults,
    state.indexes,
    state.entities,
    state.entityMeta,
    meta,
    key,
  ]);

  // If we are hard invalid we must fetch regardless of triggering or staleness
  const forceFetch = expiryStatus === ExpiryStatus.Invalid;

  const maybePromise = useMemo(() => {
    // null params mean don't do anything
    if ((Date.now() <= expiresAt && !forceFetch) || !key) return;
    // @ts-ignore
    return controller.fetch(endpoint, ...args);
    // we need to check against serialized params, since params can change frequently
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [expiresAt, controller, key, forceFetch, state.lastReset]);

  return maybePromise;
}
