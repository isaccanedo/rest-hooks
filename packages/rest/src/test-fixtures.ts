export const payload = {
  id: 5,
  title: 'hi ho',
  content: 'whatever',
  tags: ['a', 'best', 'react'],
};

export const createPayload = {
  id: 1,
  title: 'hi ho',
  content: 'whatever',
  tags: ['a', 'best', 'react'],
};

export const articlesPages = {
  prevPage: '23asdl',
  nextPage: 's3f3',
  results: [
    {
      id: 23,
      title: 'the first draft',
      content: 'the best things in life com efree',
      tags: ['one', 'two'],
    },
    {
      id: 44,
      title: 'the second book',
      content: 'the best things in life com efree',
      tags: ['hbh', 'wew'],
    },
    {
      id: 2,
      title: 'the third novel',
      content: 'the best things in life com efree',
      tags: ['free', 'honey'],
    },
    {
      id: 643,
      title: 'a long time ago',
      content: 'the best things in life com efree',
    },
  ],
};

export const users = [
  {
    id: 23,
    username: 'bob',
    email: 'bob@bob.com',
    isAdmin: false,
  },
  {
    id: 7342,
    username: 'lindsey',
    email: 'lindsey@bob.com',
    isAdmin: true,
  },
];

export const nested = [
  {
    id: 5,
    title: 'hi ho',
    content: 'whatever',
    tags: ['a', 'best', 'react'],
    author: {
      id: 23,
      username: 'bob',
    },
  },
  {
    id: 3,
    title: 'the next time',
    content: 'whatever',
    author: {
      id: 23,
      username: 'charles',
      email: 'bob@bob.com',
    },
  },
];

export const moreNested = [
  {
    id: 7,
    title: 'article 7',
    content: 'whatever',
    tags: ['blah'],
    author: {
      id: 23,
      username: 'bob',
    },
  },
  {
    id: 8,
    title: 'article 8',
    content: 'whatever',
    author: {
      id: 27,
      username: 'zac',
      email: 'zac@bob.com',
    },
  },
];

export const valuesFixture = {
  first: {
    id: 1,
    title: 'first thing',
    content: 'blah',
    tags: [],
  },
  second: {
    id: 2,
    name: 'second thing',
    content: 'blah',
    tags: [],
  },
};

export const paginatedFirstPage = {
  nextPage: '2',
  data: { results: nested },
};

export const paginatedSecondPage = {
  nextPage: null,
  data: { results: moreNested },
};
