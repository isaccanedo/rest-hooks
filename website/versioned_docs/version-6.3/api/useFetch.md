---
title: useFetch()
---

import GenericsTabs from '@site/src/components/GenericsTabs';

<GenericsTabs>

```typescript
function useFetch(
  endpoint: ReadEndpoint,
  ...args: Parameters<typeof endpoint> | [null]
): Promise<any> | undefined;
```

```typescript
function useFetch<
  E extends EndpointInterface<FetchFunction, Schema | undefined, undefined>,
  Args extends readonly [...Parameters<E>] | readonly [null],
>(endpoint: E, ...args: Args): ReturnType<E>;
```

</GenericsTabs>

Great for retrieving resources optimistically before they are needed.

This can be useful for ensuring resources early in a render tree before they are needed.

- Triggers fetch:
  - On first-render
    - or parameters change
    - or required entity is deleted
    - or imperative [invalidation](./Controller.md#invalidate) triggered
  - and When not in cache or result is considered stale
  - and When no identical requests are in flight
  - and when params are not null
- [On Error (404, 500, etc)](https://www.restapitutorial.com/httpstatuscodes.html):
  - Returned promise will reject
- On fetch returns a promise else undefined.

## Example

### Simple

```tsx
function MasterPost({ id }: { id: number }) {
  useFetch(PostResource.detail(), { id });
  // ...
}
```

### Conditional

```tsx
function MasterPost({ id, doNotFetch }: { id: number; doNotFetch: boolean }) {
  useFetch(PostResource.detail(), doNotFetch ? null : { id });
  // ...
}
```

## Useful `Endpoint`s to send

[Resource](/rest/api/resource#provided-and-overridable-methods) provides these built-in:

- detail()
- list()

Feel free to add your own [Endpoint](/rest/api/Endpoint) as well.
