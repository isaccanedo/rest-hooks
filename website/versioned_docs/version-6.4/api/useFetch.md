---
title: useFetch()
---

import GenericsTabs from '@site/src/components/GenericsTabs';

<head>
  <title>useFetch() - Declarative fetch triggers</title>
  <meta name="docsearch:pagerank" content="10"/>
</head>

<GenericsTabs>

```typescript
function useFetch(
  endpoint: ReadEndpoint,
  ...args: Parameters<typeof endpoint> | [null]
): Promise<any> | undefined;
```

```typescript
function useFetch<
  E extends EndpointInterface<FetchFunction, Schema | undefined, undefined>,
  Args extends readonly [...Parameters<E>] | readonly [null],
>(endpoint: E, ...args: Args): ReturnType<E>;
```

</GenericsTabs>

Great for retrieving resources optimistically before they are needed.

This can be useful for ensuring resources early in a render tree before they are needed.

- Triggers fetch:
  - On first-render
    - or parameters change
    - or required entity is deleted
    - or imperative [invalidation](./Controller.md#invalidate) triggered
  - and When not in cache or result is considered stale
  - and When no identical requests are in flight
  - and when params are not null
- [On Error (404, 500, etc)](https://www.restapitutorial.com/httpstatuscodes.html):
  - Returned promise will reject
- On fetch returns a promise else undefined.

:::tip

Use in combination with a data-binding hook ([useCache()](./useCache.md), [useSuspense()](./useSuspense.md), [useDLE()](./useDLE.md))
in another component.

:::

## Example

### Simple

```tsx
function MasterPost({ id }: { id: number }) {
  useFetch(PostResource.get, { id });
  // ...
}
```

### Conditional

```tsx
function MasterPost({ id, doNotFetch }: { id: number; doNotFetch: boolean }) {
  useFetch(PostResource.get, doNotFetch ? null : { id });
  // ...
}
```

## Useful `Endpoint`s to send

[Resource](/rest/api/createResource#members) provides these built-in:

- [get](/rest/api/createResource#get)
- [getList](/rest/api/createResource#getlist)

Feel free to add your own [RestEndpoint](/rest/api/RestEndpoint) as well.
