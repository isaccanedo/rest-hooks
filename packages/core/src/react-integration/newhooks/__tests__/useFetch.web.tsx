import { CoolerArticleResource, StaticArticleResource } from '__tests__/new';
import React, { Suspense } from 'react';
import { render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import nock from 'nock';

// relative imports to avoid circular dependency in tsconfig references
import {
  StateContext,
  initialState,
  State,
  ActionTypes,
  ControllerContext,
  Controller,
} from '@rest-hooks/core';

import { makeRenderRestHook, makeCacheProvider } from '../../../../../test';
import useFetch from '../useFetch';
import { users, payload } from '../test-fixtures';

async function testDispatchFetch(
  Component: React.FunctionComponent<any>,
  payloads: any[],
) {
  const dispatch = jest.fn();
  const controller = new Controller({ dispatch });

  const tree = (
    <ControllerContext.Provider value={controller}>
      <Suspense fallback={null}>
        <Component />
      </Suspense>
    </ControllerContext.Provider>
  );
  render(tree);
  expect(dispatch).toHaveBeenCalled();
  expect(dispatch.mock.calls.length).toBe(payloads.length);
  let i = 0;
  for (const call of dispatch.mock.calls) {
    delete call[0]?.meta?.createdAt;
    delete call[0]?.meta?.promise;
    expect(call[0]).toMatchSnapshot();
    const action = call[0];
    const res = await action.payload();
    expect(res).toEqual(payloads[i]);
    i++;
  }
}

function testRestHook(
  callback: () => void,
  state: State<unknown>,
  dispatch = (v: ActionTypes) => Promise.resolve(),
) {
  const controller = new Controller({ dispatch });

  return renderHook(callback, {
    wrapper: function Wrapper({ children }: { children: React.ReactNode }) {
      return (
        <ControllerContext.Provider value={controller}>
          <StateContext.Provider value={state}>
            {children}
          </StateContext.Provider>
        </ControllerContext.Provider>
      );
    },
  });
}

let mynock: nock.Scope;

beforeAll(() => {
  nock(/.*/)
    .persist()
    .defaultReplyHeaders({
      'Access-Control-Allow-Origin': '*',
      'Content-Type': 'application/json',
    })
    .options(/.*/)
    .reply(200);
  mynock = nock(/.*/).defaultReplyHeaders({
    'Access-Control-Allow-Origin': '*',
    'Content-Type': 'application/json',
  });
});

afterAll(() => {
  nock.cleanAll();
});

describe('useFetch', () => {
  let renderRestHook: ReturnType<typeof makeRenderRestHook>;
  beforeEach(() => {
    mynock.get(`/article-cooler/${payload.id}`).reply(200, payload);
    mynock.get(`/article-static/${payload.id}`).reply(200, payload);
    mynock.get(`/user/`).reply(200, users);
    renderRestHook = makeRenderRestHook(makeCacheProvider);
  });
  afterEach(() => {
    nock.cleanAll();
  });

  it('should dispatch singles', async () => {
    function FetchTester() {
      useFetch(CoolerArticleResource.get, { id: payload.id });
      return null;
    }
    await testDispatchFetch(FetchTester, [payload]);
  });

  it('should not dispatch will null params', () => {
    const dispatch = jest.fn();
    let params: any = null;
    const { rerender } = testRestHook(
      () => {
        useFetch(CoolerArticleResource.get, params);
      },
      initialState,
      dispatch,
    );
    expect(dispatch).toBeCalledTimes(0);
    params = payload;
    rerender();
    expect(dispatch).toBeCalled();
  });

  it('should dispatch with resource defined dataExpiryLength', async () => {
    function FetchTester() {
      useFetch(StaticArticleResource.get, { id: payload.id });
      return null;
    }
    await testDispatchFetch(FetchTester, [payload]);
  });

  it('should dispatch with fetch shape defined dataExpiryLength', async () => {
    function FetchTester() {
      useFetch(StaticArticleResource.longLiving, { id: payload.id });
      return null;
    }
    await testDispatchFetch(FetchTester, [payload]);
  });

  it('should dispatch with fetch shape defined errorExpiryLength', async () => {
    function FetchTester() {
      useFetch(StaticArticleResource.neverRetryOnError, { id: payload.id });
      return null;
    }
    await testDispatchFetch(FetchTester, [payload]);
  });

  it('should not refetch after expiry and render', async () => {
    let time = 1000;
    global.Date.now = jest.fn(() => time);
    nock.cleanAll();
    const fetchMock = jest.fn(() => payload);
    mynock.get(`/article-cooler/${payload.id}`).reply(200, fetchMock).persist();
    const results: any[] = [
      {
        endpoint: CoolerArticleResource.get,
        args: [{ id: payload.id }],
        response: payload,
      },
    ];
    const { result, rerender } = renderRestHook(
      () => {
        return useFetch(CoolerArticleResource.get, { id: payload.id });
      },
      { results },
    );
    await result.current;
    expect(fetchMock).toHaveBeenCalledTimes(0);
    time += 100;
    rerender();
    await result.current;
    expect(fetchMock).toHaveBeenCalledTimes(0);
    // eslint-disable-next-line require-atomic-updates
    time += 610000000;
    rerender();
    await result.current;
    rerender();
    await result.current;
    expect(fetchMock).toHaveBeenCalledTimes(0);
  });
});
