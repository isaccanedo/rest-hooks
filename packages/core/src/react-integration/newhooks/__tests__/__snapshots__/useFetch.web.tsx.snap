// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useFetch should dispatch singles 1`] = `
{
  "endpoint": [Function],
  "meta": {
    "args": [
      {
        "id": 5,
      },
    ],
    "key": "GET http://test.com/article-cooler/5",
    "options": [Function],
    "reject": [Function],
    "resolve": [Function],
    "schema": [Function],
    "throttle": true,
    "type": "read",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useFetch should dispatch with fetch shape defined dataExpiryLength 1`] = `
{
  "endpoint": [Function],
  "meta": {
    "args": [
      {
        "id": 5,
      },
    ],
    "key": "GET http://test.com/article-static/5",
    "options": [Function],
    "reject": [Function],
    "resolve": [Function],
    "schema": [Function],
    "throttle": true,
    "type": "read",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useFetch should dispatch with fetch shape defined errorExpiryLength 1`] = `
{
  "endpoint": [Function],
  "meta": {
    "args": [
      {
        "id": 5,
      },
    ],
    "key": "GET http://test.com/article-static/5",
    "options": [Function],
    "reject": [Function],
    "resolve": [Function],
    "schema": [Function],
    "throttle": true,
    "type": "read",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useFetch should dispatch with resource defined dataExpiryLength 1`] = `
{
  "endpoint": [Function],
  "meta": {
    "args": [
      {
        "id": 5,
      },
    ],
    "key": "GET http://test.com/article-static/5",
    "options": [Function],
    "reject": [Function],
    "resolve": [Function],
    "schema": [Function],
    "throttle": true,
    "type": "read",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;
